src/main/java/com/google/devtools/build/lib/packages/Attribute.java

```diff
@@ -637,14 +600,13 @@ public Builder(String name, Type<TYPE> type) {
           "analysis-test split transition");
     }
 
-    /**
-     * Defines the configuration transition for this attribute.
-     */
-    public Builder<TYPE> cfg(SplitTransitionProvider splitTransitionProvider) {
+    /** Defines the configuration transition for this attribute. */
+    // TODO(https://github.com/bazelbuild/bazel/issues/7814): Currently only for split transitions.
+    public Builder<TYPE> cfg(TransitionFactory<RuleTransitionData> splitTransitionFactory) {
       Preconditions.checkState(this.configTransition == NoTransition.INSTANCE,
           "the configuration transition is already set");
 
-      this.splitTransitionProvider = Preconditions.checkNotNull(splitTransitionProvider);
+      this.splitTransitionFactory = Preconditions.checkNotNull(splitTransitionFactory);
```

Comment:
What enforces that this `TransitionFactory` actually returns a split transition? Do we need to differentiate split and non-split `TransitionFactory` instances?

Reply:
> Looks good in principle, but I'm somewhat wary of adding unnecessary abstractions (including classes and methods). Approving on the assumption that these flags and classes will have a good reason to exist, put please keep YAGNI in mind. It's better to add abstractions when you need them instead of preemptively.

This is change 4 of N, the new abstractions will be more fully used in the next set of changes. You weren't on the original discussion with the configurability team (my apologies!) where I laid out the full plan and showed the entire (too large to review) master PR (https://github.com/bazelbuild/bazel/pull/7726).

Reply:
Added a Precondition check. This will disappear in the next set of changes, but it's a good guard in the meantime.